### PR TITLE
YJIT: Add support for passing kwargs to cfuncs

### PR DESCRIPTION
**Pull request by jhawthorn**
This is built on top of #5396, which I opened partly to test out making allocations while setting up a cfunc call.

This adds support for passing kwargs to cfuncs. This is done by calling a helper method to create the hash from the top N values on the stack (determined by the callinfo) and then moving that value onto the stack.

There are likely faster ways we could construct this hash (ex. duping an existing hash as a template and filling in the values) but I think this simple approach is fine for now and matches the behaviour of the interpreter.

In a small benchmark of GitHub's Rails app cfunc_kwargs was responsible for 17% of method call exits.

### CODE REVIEW

File: yjit_codegen.c [hunk 1]
```diff
@@ -3235,25 +3235,39 @@ c_method_tracing_currently_enabled(const jitstate_t *jit)
...
-    // If the function expects a Ruby array of arguments
-    if (cfunc->argc < 0 && cfunc->argc != -1) {
-        GEN_COUNTER_INC(cb, send_cfunc_ruby_array_varg);
-        return YJIT_CANT_COMPILE;
-    }
+    const struct rb_callinfo_kwarg *kw_arg = vm_ci_kwarg(ci);
+    const int kw_arg_num = kw_arg ? kw_arg->keyword_len : 0;
+
+    // argc adjusted for implicit kwarg hash
+    const int passed_argc = kw_arg ? argc - kw_arg_num + 1 : argc;
```
Comment (jhawthorn) [inline yjit_codegen.c:3261]: Reading through this again I don't love this name. Maybe callee_argc?

File: bootstraptest/test_yjit.rb [hunk 2]
```diff
@@ -2431,6 +2431,27 @@ def foo(required:, specified: 999, simple_default: 3, complex_default: "4".to_i)
...
+
+# variadic cfunc kwargs
+assert_equal '{:foo=>123}', %q{
+  def foo(bar)
+    bar.merge(foo: 123)
+  end
+
+  foo({})
+  foo({})
+}
```
Comment (maximecb) [inline bootstraptest/test_yjit.rb:2453]: Could also use a test case with zero arguments as that's often a corner case.

**Comment by maximecb** [comment]
Looks pretty good.

File: bootstraptest/test_yjit.rb (hunk 2 above)
Comment (jhawthorn) [reply bootstraptest/test_yjit.rb]: There isn't really a zero arguments case when passing kwargs (there's always at least one for the kwargs) but I added another non-variadic only-kwarg case and some other tests

**Review by maximecb — APPROVED** [review]
